src/translate.py

```diff
@@ -459,60 +461,71 @@ def update_min(old_or_none: Optional[int], new: int) -> int:
+            ):
+                info.subroutine_arg_top = update_min(
+                    info.subroutine_arg_top, inst.args[2].value
+                )
+
     if not info.is_leaf:
         # Compute the bounds of the callee-saved register region, including padding
-        callee_saved_words.sort()
-        callee_saved_doubles.sort()
-        bottom = callee_saved_words[0]
-
-        # Both GCC & IDO save registers in two blocks: one for word-sized registers,
-        # followed by one for doubles. The word-sized block is also padded to a
-        # multiple of 8 bytes. Check that there are no gaps in either of these blocks.
+        callee_saved_offset_and_size.sort()
+        bottom, last_size = callee_saved_offset_and_size[0]
+
+        # Both IDO & GCC save registers in two subregions:
+        # (a) One for double-sized registers
+        # (b) One for word-sized registers, padded to a multiple of 8 bytes
+        # IDO has (a) lower than (b); GCC has (b) lower than (a)
+        # Check that there are no gaps in this region, other than a single
+        # 4-byte word between subregions.
         top = bottom
-        for offset in callee_saved_words:
-            assert offset == top, DecompFailure(
-                f"Gap in callee-saved word stack region. "
-                f"Saved words: {callee_saved_words}, "
-                f"saved doubles: {callee_saved_doubles}, "
-                f"gap at: {offset} != {top}."
-            )
-            top += 4
-        # Padding: round up to multiple of 8
-        if (top % 8) == 4:
-            top += 4
-        for offset in callee_saved_doubles:
-            if offset < callee_saved_words[-1]:
-                # This offset is fake: it's a local variable; exclude it
-                continue
-            assert offset == top, DecompFailure(
-                f"Gap in callee-saved double stack region. "
-                f"Saved words: {callee_saved_words}, "
-                f"saved doubles: {callee_saved_doubles}, "
-                f"gap at: {offset} != {top}."
-            )
-            top += 8
+        internal_padding_added = False
+        for offset, size in callee_saved_offset_and_size:
+            if offset != top:
+                if (
+                    not internal_padding_added
+                    and size != last_size
+                    and offset == top + 4
+                ):
+                    top += 4
+                    internal_padding_added = True
+                else:
+                    raise DecompFailure(
+                        f"Gap in callee-saved word stack region. "
+                        f"Saved: {callee_saved_offset_and_size}, "
+                        f"gap at: {offset} != {top}."
+                    )
+            top += size
+            last_size = size
+        # Expand boundaries to multiples of 8 bytes
+        bottom -= 4 if bottom % 8 == 4 else 0
+        top += 4 if top % 8 == 4 else 0
```

Comment:
This part is surprising to me. Can't it cause sp vars to not be identified?

Reply:
As far as I can tell, the compiler always pads these subregions to 8 bytes.
This seems to match up with the different files I was looking at, and [came from this PDF](https://www.cs.unm.edu/~jeffk/cs341f09/_media/o32callingconvention.pdf):
> • All of the different sections on the stack must be double-word (8 byte) aligned. This is just so
that we can store double words onto the stack without them being unaligned.

IDO seems to put the padding low, and GCC seems to put the padding high. 

I could also just remove this pair of lines for simplicity, the 4-byte padded region is unused.

Reply:
How odd... that just seems wasteful. +1 to removing the lines, I feel like they just add risk.